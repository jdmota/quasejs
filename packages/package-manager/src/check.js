// @flow
import { read as readPkg } from "./pkg";
import {
  shouldReuse as shouldReuseLockfile,
  read as readLockfile
} from "./lockfile";

const fs = require( "fs-extra" );
const path = require( "path" );

function compare( a, b, type ) {

  for ( const name in a ) {
    if ( !b[ name ] ) {
      throw new Error( `Package.json has ${name} but lockfile has not (on ${type})` );
    }
    if ( a[ name ] !== b[ name ].savedVersion ) {
      throw new Error(
        `Package.json has ${name} with version ${a[ name ]} but lockfile has version ${b[ name ].savedVersion} (on ${type})`
      );
    }
  }

  for ( const name in b ) {
    if ( !a[ name ] ) {
      throw new Error( `Lockfile has ${name} but package.json has not (on ${type})` );
    }
  }

}

async function integrity( folder, lockfile ) {

  const checks = [];

  for ( const name in lockfile.deps ) {
    const resolvedObj = lockfile.deps[ name ];
    const idx = resolvedObj.i;
    const hash = lockfile.resolutions[ idx ][ 3 ];
    checks.push( { name, hash } );
  }

  for ( const name in lockfile.devDeps ) {
    const resolvedObj = lockfile.devDeps[ name ];
    const idx = resolvedObj.i;
    const hash = lockfile.resolutions[ idx ][ 3 ];
    checks.push( { name, hash } );
  }

  return Promise.all(
    checks.map( async( { name, hash } ) => {

      const integrityFile = await fs.readFile(
        path.join( folder, "node_modules", name, ".qpm-integrity" ),
        "utf8"
      );

      const integrity = integrityFile.trim();

      if ( integrity !== hash ) {
        throw new Error(
          `Expected integrity ${hash} but found ${integrity} for ${name}`
        );
      }

    } )
  );
}

export default async function( folder: string ) {

  const [ pkg, lockfile ] = await Promise.all( [ readPkg( folder ), readLockfile( folder ) ] );

  if ( !shouldReuseLockfile( lockfile ) ) {
    throw new Error( "Lockfile not found." );
  }

  const dependencies = pkg.dependencies || {};
  const devDependencies = pkg.devDependencies || {};

  compare( dependencies, lockfile.deps, "dependencies" );
  compare( devDependencies, lockfile.devDeps, "devDependencies" );

  await integrity( folder, lockfile );
}
