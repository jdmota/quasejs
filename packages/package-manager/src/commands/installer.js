// @flow
import reporter from "../reporters/installer";
import type { Name, Resolved, Options, Warning } from "../types";
import { mapGet } from "../utils";
import { read as readPkg, validate as validatePkg } from "../pkg";
import {
  type Lockfile,
  shouldReuse as shouldReuseLockfile,
  create as createLockfile,
  read as readLockfile,
  write as writeLockfile
} from "../lockfile";
import { ResolutionSet, Tree } from "../resolution";
import Store from "../store";
import { Resolver } from "../resolve";

const { EventEmitter } = require( "events" );

export class Installer extends EventEmitter {

  +tree: Tree;
  +rootDeps: ResolutionSet;
  +store: Store;
  +opts: Options;
  +warn: Warning => void;
  reuseLockfile: boolean;

  constructor( opts: Options ) {
    super();
    this.opts = opts;
    this.warn = ( warning: Warning ) => {
      this.emit( "warning", warning );
    };
    this.store = new Store( opts, this.warn );
    this.tree = new Tree();
    this.rootDeps = new ResolutionSet();
    this.reuseLockfile = false;
  }

  async resolution( pkg: Object, lockfile: Lockfile, newLockfile: Lockfile ) {
    this.emit( "resolutionStart" );
    const resolver = new Resolver( this, pkg, lockfile, newLockfile );
    await resolver.start();
  }

  async extraction() {

    const extractions = [];
    this.tree.forEach( resolution => {
      extractions.push( this.store.extract( resolution ).then( () => this.emit( "extractionUpdate" ) ) );
    } );

    this.emit( "extractionStart", extractions.length );

    for ( const p of extractions ) {
      await p;
    }
  }

  async linking() {

    const links = [];
    this.tree.forEach( resolution => {
      links.push( this.store.linkResolutionDeps( resolution ).then( () => this.emit( "linkingUpdate" ) ) );
    } );

    this.emit( "linkingStart", links.length );

    for ( const p of links ) {
      await p;
    }
  }

  async localLinking() {
    this.emit( "localLinkingStart", 1 );
    await this.store.linkNodeModules( this.opts.folder, this.rootDeps );
    this.emit( "localLinkingUpdate" );
  }

  async updateLockfile( newLockfile: Lockfile ) {

    this.emit( "updateLockfile" );

    const map: Map<Resolved, number> = new Map();
    this.tree.generate( newLockfile.resolutions, map );

    for ( const nameStr in newLockfile.deps ) {
      // $FlowIgnore
      const name: Name = nameStr;

      newLockfile.deps[ name ].i = mapGet( map, newLockfile.deps[ name ].resolved );
    }

    for ( const nameStr in newLockfile.devDeps ) {
      // $FlowIgnore
      const name: Name = nameStr;

      newLockfile.devDeps[ name ].i = mapGet( map, newLockfile.devDeps[ name ].resolved );
    }

    await writeLockfile( this.opts.folder, newLockfile );
  }

  async install() {

    const { opts } = this;
    const [ pkg, lockfile ] = await Promise.all( [ readPkg( opts.folder ), readLockfile( opts.folder ) ] );

    validatePkg( pkg );

    const reuseLockfile = this.reuseLockfile = !opts.update && shouldReuseLockfile( lockfile );

    this.emit( "folder", { folder: opts.folder } );
    this.emit( "lockfile", { reusing: reuseLockfile } );
    this.emit( "start" );

    const newLockfile = createLockfile();

    await this.resolution( pkg, lockfile, newLockfile );
    await this.extraction();
    await this.linking();
    await this.localLinking();
    await this.updateLockfile( newLockfile );
  }

}

export default function( opts: Options ) {
  const installer = new Installer( opts );
  reporter( installer );
  return installer.install().then( () => {
    installer.emit( "done" );
  }, err => {
    installer.emit( "error", err );
  } );
}
