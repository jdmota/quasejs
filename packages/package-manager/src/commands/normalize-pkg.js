// @flow
import { read as readPkg, write as writePkg } from "../pkg";

export default async function( folder: string ) {
  const pkg = await readPkg( folder, false, true );
  return writePkg( folder, pkg );
}
