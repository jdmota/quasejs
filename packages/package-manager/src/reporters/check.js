// @flow
import type { Checker } from "../commands/check";
import { BaseReporter } from "./base";

class CheckReporter extends BaseReporter {

  constructor() {
    super( "Looking for lockfile..." );
  }

  listen( checker: Checker ) {
    super.listen( checker );
    checker.on( "comparing", this.comparing.bind( this ) );
    checker.on( "integrity", this.warning.bind( this ) );
  }

  comparing( type: string ) {
    this.spinner.text = `Comparing ${type}...`;
  }

  integrity() {
    this.spinner.text = `Checking integrity...`;
  }

}

export default function( checker: Checker ) {
  new CheckReporter().listen( checker );
}
