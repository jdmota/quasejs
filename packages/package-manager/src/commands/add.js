// @flow
import type { Options } from "../types";
import { error } from "../utils";
import { read, readGlobal, write, add, normalizeType } from "../pkg";
import installer from "./installer";

// $FlowIgnore
const latestVersion = require( "latest-version" );
const semver = require( "semver" );

async function inputMapper( required: string ): Promise<{ name: string, version: string }> {
  const i = required.lastIndexOf( "@" );

  if ( i <= 0 ) {
    const name = required.trim();

    return {
      name,
      version: await latestVersion( name )
    };
  }

  const name = required.slice( 0, i ).trim();
  const version = required.slice( i + 1 ).trim();

  const validatedRange = semver.validRange( version );

  if ( validatedRange == null ) {
    throw error( `Invalid version/range for ${name}` );
  }

  return {
    name,
    version: validatedRange
  };
}

/* eslint-disable no-console */

export default async function( options: Options, input: string[] ) {
  const pkg = options.global ? await readGlobal( options.folder ) : await read( options.folder );
  const packages = await Promise.all( input.map( inputMapper ) );
  const added = add( pkg, packages, normalizeType( options.type ) );
  if ( added.length > 0 ) {
    await write( options.folder, pkg );

    console.log( "Adding:" );
    for ( const { name, version } of added ) {
      console.log( `  ${name}@${version}` );
    }
    console.log( "" );
  } else {
    throw error( "Nothing was added" );
  }
  return installer( options );
}
