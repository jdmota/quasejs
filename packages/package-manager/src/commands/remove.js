// @flow
import type { Options } from "../types";
import { error } from "../utils";
import { read, readGlobal, write, remove, normalizeType } from "../pkg";
import installer from "./installer";

/* eslint-disable no-console */

const logSymbols = require( "log-symbols" );

export default async function( options: Options, input: string[] ) {
  const pkg = options.global ? await readGlobal( options.folder ) : await read( options.folder );
  const removed = remove( pkg, input, normalizeType( options.type ) );
  if ( removed.length > 0 ) {
    await write( options.folder, pkg );

    console.log( `${logSymbols.info} Removing:` );
    for ( const { name, version } of removed ) {
      console.log( `  - ${name}@${version}` );
    }
    console.log( "" );
  } else {
    throw error( "Nothing was removed" );
  }
  return installer( options );
}
