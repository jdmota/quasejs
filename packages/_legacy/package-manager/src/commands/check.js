// @flow
import { CheckReporter } from "../reporters/check";
import { type AliasName, toStr } from "../types";
import { error, readJSON } from "../utils";
import { read as readPkg } from "../pkg";
import { shouldReuse as shouldReuseLockfile, read as readLockfile } from "../lockfile";

const path = require( "path" );

function compare( a, b, type ) {

  for ( const nameStr in a ) {
    // $FlowIgnore
    const name: AliasName = nameStr;

    if ( !b[ name ] ) {
      throw error( `Package.json has ${nameStr} but lockfile has not (on ${type}).` );
    }
    if ( a[ name ] !== b[ name ].spec ) {
      throw error(
        `Package.json has ${nameStr} with version ${a[ name ]} but lockfile has version ${toStr( b[ name ].spec )} (on ${type}).`
      );
    }
  }

  for ( const nameStr in b ) {
    // $FlowIgnore
    const name: AliasName = nameStr;

    if ( !a[ name ] ) {
      throw error( `Lockfile has ${nameStr} but package.json has not (on ${type}).` );
    }
  }

}

async function integrity( folder, lockfile ) {

  const checks = [];

  for ( const nameStr in lockfile.deps ) {
    // $FlowIgnore
    const alias: AliasName = nameStr;

    const resolvedObj = lockfile.deps[ alias ];
    const idx = resolvedObj.i;
    const name = lockfile.resolutions[ idx ][ 0 ];
    const version = lockfile.resolutions[ idx ][ 1 ];
    const hash = lockfile.resolutions[ idx ][ 3 ];
    checks.push( { pkg: `${toStr( name )}@${toStr( version )}`, alias, hash } );
  }

  for ( const nameStr in lockfile.devDeps ) {
    // $FlowIgnore
    const alias: AliasName = nameStr;

    const resolvedObj = lockfile.devDeps[ alias ];
    const idx = resolvedObj.i;
    const name = lockfile.resolutions[ idx ][ 0 ];
    const version = lockfile.resolutions[ idx ][ 1 ];
    const hash = lockfile.resolutions[ idx ][ 3 ];
    checks.push( { pkg: `${toStr( name )}@${toStr( version )}`, alias, hash } );
  }

  return Promise.all(
    checks.map( async( { pkg, alias, hash } ) => {

      // TODO check folder integrity?

      const { integrity } = await readJSON(
        path.join( folder, "node_modules", alias, ".qpm" )
      );

      if ( integrity === undefined ) {
        throw error( `It seems that ${pkg} is not installed.` );
      }

      if ( integrity !== hash ) {
        throw error(
          `Expected integrity ${toStr( hash )} but found ${integrity} for ${pkg}.`
        );
      }

    } )
  );
}

export class Checker {

  +reporter: CheckReporter;

  constructor( reporter: CheckReporter ) {
    this.reporter = reporter;
  }

  async check( folder: string ) {

    this.reporter.start();

    const [ pkg, lockfile ] = await Promise.all( [ readPkg( folder ), readLockfile( folder ) ] );

    if ( !shouldReuseLockfile( lockfile ) ) {
      this.reporter.warning( {
        code: "LOCKFILE_NOT_FOUND",
        message: "Lockfile not found."
      } );
      return;
    }

    const dependencies = pkg.dependencies || {};
    const devDependencies = pkg.devDependencies || {};

    this.reporter.comparing( "dependencies" );

    compare( dependencies, lockfile.deps, "dependencies" );

    this.reporter.comparing( "devDependencies" );

    compare( devDependencies, lockfile.devDeps, "devDependencies" );

    this.reporter.integrity();

    await integrity( folder, lockfile );
  }

}

export default function( folder: string ) {
  const reporter = new CheckReporter();
  const checker = new Checker( reporter );
  return checker.check( folder ).then( () => {
    reporter.done();
  }, err => {
    reporter.error( err );
  } );
}
