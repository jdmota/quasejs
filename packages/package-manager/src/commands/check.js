// @flow
import { CheckReporter } from "../reporters/check";
import { type Name, toStr } from "../types";
import { error, readJSON } from "../utils";
import { read as readPkg } from "../pkg";
import { shouldReuse as shouldReuseLockfile, read as readLockfile } from "../lockfile";

const path = require( "path" );

function compare( a, b, type ) {

  for ( const nameStr in a ) {
    // $FlowIgnore
    const name: Name = nameStr;

    if ( !b[ name ] ) {
      throw error( `Package.json has ${nameStr} but lockfile has not (on ${type}).` );
    }
    if ( a[ name ] !== b[ name ].savedVersion ) {
      throw error(
        `Package.json has ${nameStr} with version ${a[ name ]} but lockfile has version ${toStr( b[ name ].savedVersion )} (on ${type}).`
      );
    }
  }

  for ( const nameStr in b ) {
    // $FlowIgnore
    const name: Name = nameStr;

    if ( !a[ name ] ) {
      throw error( `Lockfile has ${nameStr} but package.json has not (on ${type}).` );
    }
  }

}

async function integrity( folder, lockfile ) {

  const checks = [];

  for ( const nameStr in lockfile.deps ) {
    // $FlowIgnore
    const name: Name = nameStr;

    const resolvedObj = lockfile.deps[ name ];
    const idx = resolvedObj.i;
    const version = lockfile.resolutions[ idx ][ 1 ];
    const hash = lockfile.resolutions[ idx ][ 3 ];
    checks.push( { pkg: `${nameStr}@${toStr( version )}`, name, hash } );
  }

  for ( const nameStr in lockfile.devDeps ) {
    // $FlowIgnore
    const name: Name = nameStr;

    const resolvedObj = lockfile.devDeps[ name ];
    const idx = resolvedObj.i;
    const version = lockfile.resolutions[ idx ][ 1 ];
    const hash = lockfile.resolutions[ idx ][ 3 ];
    checks.push( { pkg: `${nameStr}@${toStr( version )}`, name, hash } );
  }

  return Promise.all(
    checks.map( async( { pkg, name, hash } ) => {

      // TODO check folder integrity?

      const { integrity } = await readJSON(
        path.join( folder, "node_modules", name, ".qpm" )
      );

      if ( integrity === undefined ) {
        throw error( `It seems that ${pkg} is not installed.` );
      }

      if ( integrity !== hash ) {
        throw error(
          `Expected integrity ${toStr( hash )} but found ${integrity} for ${pkg}.`
        );
      }

    } )
  );
}

export class Checker {

  +reporter: CheckReporter;

  constructor( reporter: CheckReporter ) {
    this.reporter = reporter;
  }

  async check( folder: string ) {

    this.reporter.start();

    const [ pkg, lockfile ] = await Promise.all( [ readPkg( folder ), readLockfile( folder ) ] );

    if ( !shouldReuseLockfile( lockfile ) ) {
      this.reporter.warning( {
        code: "LOCKFILE_NOT_FOUND",
        message: "Lockfile not found."
      } );
      return;
    }

    const dependencies = pkg.dependencies || {};
    const devDependencies = pkg.devDependencies || {};

    this.reporter.comparing( "dependencies" );

    compare( dependencies, lockfile.deps, "dependencies" );

    this.reporter.comparing( "devDependencies" );

    compare( devDependencies, lockfile.devDeps, "devDependencies" );

    this.reporter.integrity();

    await integrity( folder, lockfile );
  }

}

export default function( folder: string ) {
  const reporter = new CheckReporter();
  const checker = new Checker( reporter );
  return checker.check( folder ).then( () => {
    reporter.done();
  }, err => {
    reporter.error( err );
  } );
}
