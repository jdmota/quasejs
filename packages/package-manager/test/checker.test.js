import { Checker } from "../src/commands/check";

const path = require( "path" );

it( "duplicate dependencies in package.json", async() => {

  await expect(
    new Checker().check( path.resolve( __dirname, "test-folders/duplicate-dep" ) )
  ).rejects.toThrowErrorMatchingSnapshot();

} );
