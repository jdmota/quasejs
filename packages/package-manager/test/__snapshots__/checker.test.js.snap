// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`duplicate dependencies in package.json 1`] = `"read-pkg appears in both dependencies and devDependencies"`;
